File: fundamentals-of-software-architecture-for-big-data/simple-aged-cache/src/main/java/io/collective/ExpirableEntry.java
package io.collective;

import java.time.Clock;

class ExpirableEntry {

  private Object key;
  private Object value;

  private long expiresAt;

  public ExpirableEntry next = null;

  protected ExpirableEntry(Object key, Object value, long expiresAt) {
      this.key = key;
      this.value = value;
      this.expiresAt = expiresAt;
  }

  public boolean isValid(Clock clock) {
      return clock.millis() <= expiresAt;
  }

  public boolean hasKey(Object key) {
      return this.key.equals(key);
  }

  public Object getValue() {
      return value;
  }

}

File: fundamentals-of-software-architecture-for-big-data/simple-aged-cache/bin/main/io/collective/ExpirableEntry.kt
package io.collective

import java.time.Clock

class ExpirableEntryK {

  val key: Any?
  val value: Any?
  val expiresAt: Long

  var next: ExpirableEntryK? = null 

  constructor(
      key: Any?, 
      value: Any?, 
      expiresAt: Long
  ) {
      this.key = key       
      this.value = value
      this.expiresAt = expiresAt
  }    

  fun isValid(clock: Clock): Boolean {
      return clock.millis() <= expiresAt;
  }
}